### PR TITLE
Use v0.14 api in groonga

### PR DESCRIPTION
**Pull request by kenhys**

### CODE REVIEW

**Comment by cosmo0920** [comment]
Is there a plan to add descriptions to use v0.14 style workers in this PR?
Documentation is out of scope in this PR?

**Comment by kenhys** [comment]
Yes. Documentation is out of scope in THIS PR.

**Comment by kou** [comment]
Documentation should include THIS PR. Because we can't migrate to new API without we know how to use new workers style.

**Comment by kenhys** [comment]
I've checked buffered behavior as follows.

1. launch fluentd without Groonga.
2. send request to fluentd (failed log as expected)
3. launch master and slave Groonga.
4. send kill -1 to fluentd supervisor for restarting fluentd. 
5. check master and slave tables whether failed request is flushed to them.

# step 1

```
2017-04-13 11:22:36 +0900 [info]: reading config file path="fluent.conf"
2017-04-13 11:22:36 +0900 [info]: starting fluentd-0.14.14 pid=5870
2017-04-13 11:22:36 +0900 [info]: spawn command to main:  cmdline=["/home/kenhys/.rvm/rubies/ruby-2.4.1/bin/ruby", "-Eascii-8bit:ascii-8bit", "/home/kenhys/.rvm/gems/ruby-2.4.1@fluentd-latest/bin/fluentd", "-c", "fluent.conf", "--under-supervisor"]
2017-04-13 11:22:36 +0900 [info]: gem 'fluent-plugin-groonga' version '1.1.7.1'
2017-04-13 11:22:36 +0900 [info]: gem 'fluentd' version '0.14.14'
2017-04-13 11:22:36 +0900 [info]: adding match pattern="groonga.command.*" type="groonga"
2017-04-13 11:22:36 +0900 [info]: #0 'flush_interval' is configured at out side of <buffer>. 'flush_mode' is set to 'interval' to keep existing behaviour
2017-04-13 11:22:36 +0900 [info]: adding source type="groonga"
2017-04-13 11:22:36 +0900 [info]: using configuration file: <ROOT>
  <system>
    workers 1
  </system>
  <source>
    @type groonga
    protocol http
    bind "127.0.0.1"
    port 10041
    real_host "127.0.0.1"
    real_port 20041
  </source>
  <match groonga.command.*>
    @type groonga
    protocol http
    host "localhost"
    port 20042
    flush_interval 1s
    <buffer>
      flush_interval 1s
    </buffer>
  </match>
</ROOT>
2017-04-13 11:22:36 +0900 [info]: #0 starting fluentd worker pid=5876 ppid=5870 worker=0
2017-04-13 11:22:36 +0900 [info]: #0 fluentd worker is now running worker=0
2017-04-13 11:28:26 +0900 [error]: #0 [input][groonga][connect][error] failed to connect to Groonga: real_host="127.0.0.1" real_port=20041
```

# step 2

send request to fluentd.

```
 curl http://localhost:10041/d/table_create?name=plugin_test2&flags=TABLE_HASH_KEY&key_type=ShortText  
curl: (56) Recv failure: 接続が相手からリセットされました     
```

# step 3

launch master and slave Groonga

```
% ps -ef G groonga
kenhys   10289     1  0 11:33 pts/12   00:00:00 groonga --protocol http --port 20041 -d testdb/master.db
kenhys   10379     1  0 11:33 pts/12   00:00:00 groonga --protocol http --port 20042 -d testdb/slave.db
```
# step 4

restart fluentd.

```
kill -1 5870
```

```
2017-04-13 11:28:26 +0900 [error]: #0 [input][groonga][connect][error] failed to connect to Groonga: real_host="127.0.0.1" real_port=20041
w2017-04-13 11:34:19 +0900 [info]: #0 fluentd worker is now stopping worker=0
2017-04-13 11:34:19 +0900 [info]: #0 shutting down fluentd worker worker=0
2017-04-13 11:34:19 +0900 [info]: #0 shutting down input plugin type=:groonga plugin_id="object:2ac927261054"
2017-04-13 11:34:19 +0900 [info]: #0 shutting down output plugin type=:groonga plugin_id="object:2ac9277ddc44"
2017-04-13 11:34:19 +0900 [info]: Worker 0 finished unexpectedly with status 0
2017-04-13 11:34:19 +0900 [info]: gem 'fluent-plugin-groonga' version '1.1.7.1'
2017-04-13 11:34:19 +0900 [info]: gem 'fluentd' version '0.14.14'
2017-04-13 11:34:19 +0900 [info]: adding match pattern="groonga.command.*" type="groonga"
2017-04-13 11:34:19 +0900 [info]: #0 'flush_interval' is configured at out side of <buffer>. 'flush_mode' is set to 'interval' to keep existing behaviour
2017-04-13 11:34:20 +0900 [info]: adding source type="groonga"
2017-04-13 11:34:20 +0900 [info]: using configuration file: <ROOT>
  <system>
    workers 1
  </system>
  <source>
    @type groonga
    protocol http
    bind "127.0.0.1"
    port 10041
    real_host "127.0.0.1"
    real_port 20041
  </source>
  <match groonga.command.*>
    @type groonga
    protocol http
    host "localhost"
    port 20042
    flush_interval 1s
    <buffer>
      flush_interval 1s
    </buffer>
  </match>
</ROOT>
2017-04-13 11:34:20 +0900 [info]: #0 starting fluentd worker pid=10999 ppid=5870 worker=0
2017-04-13 11:34:20 +0900 [info]: #0 fluentd worker is now running worker=0
```

# step 5

check buffer is flushed and expected table is created both of them.

```
% ./create.sh m
+ echo Checking master
Checking master
+ curl http://localhost:20041/d/table_list
[[0,1492051095.225338,0.0002312660217285156],[[["id","UInt32"],["name","ShortText"],["path","ShortText"],["flags","ShortText"],["domain","ShortText"],["range","ShortText"],["default_tokenizer","ShortText"],["normalizer","ShortText"]]]]
% ./create.sh s
+ echo Checking slave
Checking slave
+ curl http://localhost:20042/d/table_list
[[0,1492051096.240427,0.0001485347747802734],[[["id","UInt32"],["name","ShortText"],["path","ShortText"],["flags","ShortText"],["domain","ShortText"],["range","ShortText"],["default_tokenizer","ShortText"],["normalizer","ShortText"]]]]
```

**Comment by kou** [comment]
Buffer is for output plugin not input plugin. Did you want to comment it to #16?

BYW1: `flush_interval` at `<match>` level should be removed. We can't confirm which `flush_interval` in `<match>` or `<buffer>` is used.

```text
  <match groonga.command.*>
    @type groonga
    protocol http
    host "localhost"
    port 20042
    flush_interval 1s
    <buffer>
      flush_interval 1s
    </buffer>
  </match>
```

BTW2: There is a warning(?) message in log:

```text
2017-04-13 11:34:19 +0900 [info]: #0 'flush_interval' is configured at out side of <buffer>. 'flush_mode' is set to 'interval' to keep existing behaviour
```

BTW3: Why did your restart `fluentd` instead of waiting 1 sec (the specified value by `flush_interval`)? Restarting is related with `flush_interval`?

**Comment by kenhys** [comment]
Oops, I intended to comment against #16.

**Comment by kenhys** [comment]
Added documentation.

**Comment by kenhys** [comment]
Migrate from old style to v0.14 is now working on. via https://github.com/groonga/fluent-plugin-groonga/pull/17

**Comment by cosmo0920** [comment]
We should override `multi_workers_ready?` to return true.
Otherwise, Fluentd v0.14 forbid to work with multiple workers.

File: lib/fluent/plugin/in_groonga.rb [hunk 1]
```diff
@@ -132,7 +131,6 @@ def configure(conf)
 
       def start
         listen_socket = TCPServer.new(@bind, @port)
```
Comment (cosmo0920) [inline lib/fluent/plugin/in_groonga.rb]: For **minimal** multiple workers support, I guess that we can use `server_create_tcp_socket(shared, bind, port)` helper method in server plugin helper.

**Comment by cosmo0920** [comment]
An I've noticed that we should override `multi_workers_ready?` to return true in `out_groonga` to support multiple workers.

File: lib/fluent/plugin/in_groonga.rb (hunk 1 above)
Comment (cosmo0920) [inline lib/fluent/plugin/in_groonga.rb]: Any progress?

Comment (kou) [inline lib/fluent/plugin/in_groonga.rb]: I'll implement it today.

Comment (cosmo0920) [inline lib/fluent/plugin/in_groonga.rb]: I've tried to work with multi worker environment with the following patches and confirm to pass test cases:
https://github.com/groonga/fluent-plugin-groonga/compare/use-v0.14-api-in-groonga...cosmo0920:adopting-multi-workers

**Comment by kou** [comment]
@kenhys Could you extract document change as a new pull request?

**Comment by kenhys** [comment]
OK, I'll do it.

**Comment by kenhys** [comment]
This PR is obsoleted. close.

**Pull request closed without merging**